rust/src/io/commit/dynamodb.rs

```diff
@@ -0,0 +1,530 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! DynamoDB based external manifest store
+//!
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use aws_sdk_dynamodb::error::SdkError;
+use aws_sdk_dynamodb::operation::get_item::builders::GetItemFluentBuilder;
+use aws_sdk_dynamodb::operation::put_item::builders::PutItemFluentBuilder;
+use aws_sdk_dynamodb::operation::query::builders::QueryFluentBuilder;
+use aws_sdk_dynamodb::types::{AttributeValue, KeyType};
+use aws_sdk_dynamodb::Client;
+use snafu::OptionExt;
+
+use crate::error::{IOSnafu, NotFoundSnafu};
+use crate::io::commit::external_manifest::ExternalManifestStore;
+use crate::{Error, Result};
+
+impl<E> From<SdkError<E>> for Error {
+    fn from(e: SdkError<E>) -> Self {
+        Self::IO {
+            message: format!("dynamodb error: {e}"),
+        }
+    }
+}
+
+/// An external manifest store backed by DynamoDB
+///
+/// When calling DynamoDBExternalManifestStore::new_external_store()
+/// the key schema, (PK, SK), is checked. If the table does not exist,
+/// or the key schema is not as expected, an error is returned.
+///
+/// The table schema is expected as follows:
+/// PK: base_uri -- string
+/// SK: version -- number
+/// path -- string
+/// commiter -- string
+///
+/// Consistency: This store is expected to have read-after-write consistency
+/// consistent_read should always be set to true
+///
+/// Transaction Safty: This store uses DynamoDB conditional write to ensure
+/// only one writer can win per version.
+#[derive(Debug)]
+struct DynamoDBExternalManifestStore {
```

Comment:
Does this PR expose this in any way? (It's private right now, right?) Is that for a future PR?

Reply:
I will expose this in the next PR, where we parse for `s3+ddb://` scheme and construct `DynamoDBExternalManifestStore` as a `CommitHandler`